From 2c2a46942f1af43cf887079e974857f5913ea520 Mon Sep 17 00:00:00 2001
From: huawei <huawei@huawei.com>
Date: Tue, 16 Nov 2021 17:51:51 +0800
Subject: [PATCH openEuler-1.0-LTS 1/3] EMMC: Ascend hisi customized emmc host

--------------------

To identify Ascend HiSilicon emmc chip, we add a customized property
to dts. In this patch we add an interface to read this property. At
the same time, we provided a switch, which is CONFIG_ASCEND_HISI_MMC,
for you to get rid of our modifications.

To support Ascend HiSilicon emmc chip, we add specific extensions in
mmc core, include mmc card sleep/awake/reset, cache ctrl, power vcc
adaptions and etc. Also, we enable some features for emmc5.0/5.1.

Modified mmc ops for Ascend HiSilicon emmc card because of the
difference bwteen our emmc card and others. we used customized
property in dts to judge if it is our chips. In this patch, we
modified emmc card init process, rate setting and ext_csd register
decode/encode. Modified sd suspend, resume, reset and init process
for ascend HiSilicon chip.

Signed-off-by: huawei <huawei@huawei.com>
---
 drivers/mmc/Kconfig              |  10 +
 drivers/mmc/core/Makefile        |   1 +
 drivers/mmc/core/block.c         |   3 +
 drivers/mmc/core/bus.c           |  11 +
 drivers/mmc/core/core.c          |  65 ++-
 drivers/mmc/core/core.h          |  18 +
 drivers/mmc/core/hisi_core_mmc.c | 818 +++++++++++++++++++++++++++++++
 drivers/mmc/core/hisi_core_mmc.h |  52 ++
 drivers/mmc/core/host.c          |  45 +-
 drivers/mmc/core/mmc.c           | 395 ++++++++++++++-
 drivers/mmc/core/sd.c            |  83 +++-
 include/linux/mmc/card.h         | 105 ++++
 include/linux/mmc/core.h         |   4 +-
 include/linux/mmc/host.h         | 116 +++++
 include/linux/mmc/mmc.h          |  46 ++
 include/linux/mmc/pm.h           |   1 +
 16 files changed, 1729 insertions(+), 44 deletions(-)
 mode change 100644 => 100755 drivers/mmc/Kconfig
 mode change 100644 => 100755 drivers/mmc/core/Makefile
 mode change 100644 => 100755 drivers/mmc/core/block.c
 mode change 100644 => 100755 drivers/mmc/core/bus.c
 mode change 100644 => 100755 drivers/mmc/core/core.c
 mode change 100644 => 100755 drivers/mmc/core/core.h
 create mode 100755 drivers/mmc/core/hisi_core_mmc.c
 create mode 100755 drivers/mmc/core/hisi_core_mmc.h
 mode change 100644 => 100755 drivers/mmc/core/host.c
 mode change 100644 => 100755 drivers/mmc/core/mmc.c
 mode change 100644 => 100755 drivers/mmc/core/sd.c
 mode change 100644 => 100755 include/linux/mmc/card.h
 mode change 100644 => 100755 include/linux/mmc/core.h
 mode change 100644 => 100755 include/linux/mmc/host.h
 mode change 100644 => 100755 include/linux/mmc/mmc.h
 mode change 100644 => 100755 include/linux/mmc/pm.h

diff --git a/drivers/mmc/Kconfig b/drivers/mmc/Kconfig
old mode 100644
new mode 100755
index ec2138831..8b29ecadd
--- a/drivers/mmc/Kconfig
+++ b/drivers/mmc/Kconfig
@@ -12,6 +12,16 @@ menuconfig MMC
 	  If you want MMC/SD/SDIO support, you should say Y here and
 	  also to your specific host controller driver.

+config ASCEND_HISI_MMC
+	bool "Ascend HiSilicon MMC card support"
+	depends on MMC
+	default n
+	help
+	  This selects for Hisilicon SoC specific extensions to the
+	  Synopsys DesignWare Memory Card Interface driver.
+	  You should select this option if you want mmc support on
+	  Ascend platform.
+
 if MMC

 source "drivers/mmc/core/Kconfig"
diff --git a/drivers/mmc/core/Makefile b/drivers/mmc/core/Makefile
old mode 100644
new mode 100755
index abba078f7..5d76782ff
--- a/drivers/mmc/core/Makefile
+++ b/drivers/mmc/core/Makefile
@@ -18,3 +18,4 @@ obj-$(CONFIG_MMC_BLOCK)		+= mmc_block.o
 mmc_block-objs			:= block.o queue.o
 obj-$(CONFIG_MMC_TEST)		+= mmc_test.o
 obj-$(CONFIG_SDIO_UART)		+= sdio_uart.o
+mmc_core-$(CONFIG_ASCEND_HISI_MMC)	+= hisi_core_mmc.o
diff --git a/drivers/mmc/core/block.c b/drivers/mmc/core/block.c
old mode 100644
new mode 100755
index 60eac66dc..d55f3abfc
--- a/drivers/mmc/core/block.c
+++ b/drivers/mmc/core/block.c
@@ -2247,6 +2247,9 @@ enum mmc_issued mmc_blk_mq_issue_rq(struct mmc_queue *mq, struct request *req)

 static inline int mmc_blk_readonly(struct mmc_card *card)
 {
+	if (mmc_is_ascend_customized(card->host->parent))
+		return 0;
+
 	return mmc_card_readonly(card) ||
 	       !(card->csd.cmdclass & CCC_BLOCK_WRITE);
 }
diff --git a/drivers/mmc/core/bus.c b/drivers/mmc/core/bus.c
old mode 100644
new mode 100755
index fc92c6c1c..a12d26018
--- a/drivers/mmc/core/bus.c
+++ b/drivers/mmc/core/bus.c
@@ -134,6 +134,17 @@ static void mmc_bus_shutdown(struct device *dev)
 	struct mmc_host *host = card->host;
 	int ret;

+	if (mmc_is_ascend_customized(host->parent)) {
+		int present = 1;
+
+		host->rescan_disable = 1;
+		if (host->ops->get_cd)
+			present = host->ops->get_cd(host);
+
+		if (present == 0)
+			return;
+	}
+
 	if (dev->driver && drv->shutdown)
 		drv->shutdown(card);

diff --git a/drivers/mmc/core/core.c b/drivers/mmc/core/core.c
old mode 100644
new mode 100755
index 0a74785e5..c5c021bdc
--- a/drivers/mmc/core/core.c
+++ b/drivers/mmc/core/core.c
@@ -36,8 +36,12 @@
 #include <linux/mmc/sd.h>
 #include <linux/mmc/slot-gpio.h>

+#include <linux/jiffies.h>
+
 #define CREATE_TRACE_POINTS
 #include <trace/events/mmc.h>
+#include <linux/version.h>
+

 #include "core.h"
 #include "card.h"
@@ -924,7 +928,7 @@ EXPORT_SYMBOL(mmc_put_card);
  * Internal function that does the actual ios call to the host driver,
  * optionally printing some debug output.
  */
-static inline void mmc_set_ios(struct mmc_host *host)
+void mmc_set_ios(struct mmc_host *host)
 {
 	struct mmc_ios *ios = &host->ios;

@@ -1021,7 +1025,8 @@ void mmc_set_initial_state(struct mmc_host *host)
 		host->ios.chip_select = MMC_CS_HIGH;
 	else
 		host->ios.chip_select = MMC_CS_DONTCARE;
-	host->ios.bus_mode = MMC_BUSMODE_PUSHPULL;
+	if (!mmc_is_ascend_customized(host->parent))
+		host->ios.bus_mode = MMC_BUSMODE_PUSHPULL;
 	host->ios.bus_width = MMC_BUS_WIDTH_1;
 	host->ios.timing = MMC_TIMING_LEGACY;
 	host->ios.drv_type = 0;
@@ -1533,7 +1538,15 @@ int mmc_set_uhs_voltage(struct mmc_host *host, u32 ocr)

 	cmd.opcode = SD_SWITCH_VOLTAGE;
 	cmd.arg = 0;
-	cmd.flags = MMC_RSP_R1 | MMC_CMD_AC;
+	if (mmc_is_ascend_customized(host->parent))
+		/*
+		 * Here for using Kingsdon SD produced after 03-2015,
+		 * We set long resp and delete CRC flag for cmd11.
+		 */
+		cmd.flags = MMC_RSP_PRESENT | MMC_RSP_OPCODE |
+			    MMC_CMD_AC | MMC_RSP_136;
+	else
+		cmd.flags = MMC_RSP_R1 | MMC_CMD_AC;

 	err = mmc_wait_for_cmd(host, &cmd, 0);
 	if (err)
@@ -1651,6 +1664,8 @@ void mmc_power_up(struct mmc_host *host, u32 ocr)
 	mmc_pwrseq_pre_power_on(host);

 	host->ios.vdd = fls(ocr) - 1;
+	if (mmc_is_ascend_customized(host->parent))
+		host->ios.bus_mode = MMC_BUSMODE_PUSHPULL;
 	host->ios.power_mode = MMC_POWER_UP;
 	/* Set initial state and call mmc_set_ios */
 	mmc_set_initial_state(host);
@@ -1687,6 +1702,9 @@ void mmc_power_off(struct mmc_host *host)
 	host->ios.clock = 0;
 	host->ios.vdd = 0;

+	if (!mmc_host_is_spi(host) && mmc_is_ascend_customized(host->parent))
+		host->ios.bus_mode = MMC_BUSMODE_OPENDRAIN;
+
 	host->ios.power_mode = MMC_POWER_OFF;
 	/* Set initial state and call mmc_set_ios */
 	mmc_set_initial_state(host);
@@ -1720,7 +1738,7 @@ static void __mmc_release_bus(struct mmc_host *host)
 /*
  * Increase reference count of bus operator
  */
-static inline void mmc_bus_get(struct mmc_host *host)
+void mmc_bus_get(struct mmc_host *host)
 {
 	unsigned long flags;

@@ -1733,7 +1751,7 @@ static inline void mmc_bus_get(struct mmc_host *host)
  * Decrease reference count of bus operator and free it if
  * it is the last reference.
  */
-static inline void mmc_bus_put(struct mmc_host *host)
+void mmc_bus_put(struct mmc_host *host)
 {
 	unsigned long flags;

@@ -2515,9 +2533,10 @@ static int mmc_rescan_try_freq(struct mmc_host *host, unsigned freq)
 		mmc_send_if_cond(host, host->ocr_avail);

 	/* Order's important: probe SDIO, then SD, then MMC */
-	if (!(host->caps2 & MMC_CAP2_NO_SDIO))
-		if (!mmc_attach_sdio(host))
-			return 0;
+	if (!mmc_is_ascend_customized(host->parent))
+		if (!(host->caps2 & MMC_CAP2_NO_SDIO))
+			if (!mmc_attach_sdio(host))
+				return 0;

 	if (!(host->caps2 & MMC_CAP2_NO_SD))
 		if (!mmc_attach_sd(host))
@@ -2604,12 +2623,18 @@ void mmc_rescan(struct work_struct *work)
 		container_of(work, struct mmc_host, detect.work);
 	int i;

-	if (host->rescan_disable)
+	if (host->rescan_disable) {
+		if (mmc_is_ascend_customized(host->parent))
+			goto out;
 		return;
+	}

 	/* If there is a non-removable card registered, only scan once */
-	if (!mmc_card_is_removable(host) && host->rescan_entered)
+	if (!mmc_card_is_removable(host) && host->rescan_entered) {
+		if (mmc_is_ascend_customized(host->parent))
+			goto out;
 		return;
+	}
 	host->rescan_entered = 1;

 	if (host->trigger_card_event && host->ops->card_event) {
@@ -2660,6 +2685,13 @@ void mmc_rescan(struct work_struct *work)
 	for (i = 0; i < ARRAY_SIZE(freqs); i++) {
 		if (!mmc_rescan_try_freq(host, max(freqs[i], host->f_min)))
 			break;
+		else if (mmc_is_ascend_customized(host->parent)) {
+			if (host->index == 0) {
+				pr_err("%s emmc init failed,\n", __func__);
+				pr_err("need to reinit\n");
+				continue;
+			}
+		}
 		if (freqs[i] <= host->f_min)
 			break;
 	}
@@ -2681,17 +2713,18 @@ void mmc_start_host(struct mmc_host *host)
 		mmc_power_up(host, host->ocr_avail);
 		mmc_release_host(host);
 	}
-
-	mmc_gpiod_request_cd_irq(host);
+	if (!mmc_is_ascend_customized(host->parent))
+		mmc_gpiod_request_cd_irq(host);
 	_mmc_detect_change(host, 0, false);
 }

 void mmc_stop_host(struct mmc_host *host)
 {
-	if (host->slot.cd_irq >= 0) {
-		mmc_gpio_set_cd_wake(host, false);
-		disable_irq(host->slot.cd_irq);
-	}
+	if (!mmc_is_ascend_customized(host->parent))
+		if (host->slot.cd_irq >= 0) {
+			mmc_gpio_set_cd_wake(host, false);
+			disable_irq(host->slot.cd_irq);
+		}

 	host->rescan_disable = 1;
 	cancel_delayed_work_sync(&host->detect);
diff --git a/drivers/mmc/core/core.h b/drivers/mmc/core/core.h
old mode 100644
new mode 100755
index 087ba68b2..3c3f0ca80
--- a/drivers/mmc/core/core.h
+++ b/drivers/mmc/core/core.h
@@ -32,6 +32,13 @@ struct mmc_bus_ops {
 	int (*shutdown)(struct mmc_host *);
 	int (*hw_reset)(struct mmc_host *);
 	int (*sw_reset)(struct mmc_host *);
+#ifdef CONFIG_ASCEND_HISI_MMC
+	int (*awake)(struct mmc_host *host);
+	int (*sleep)(struct mmc_host *host);
+	int (*power_save)(struct mmc_host *host);
+	int (*power_restore)(struct mmc_host *host);
+	int (*reinit_card)(struct mmc_host *host);
+#endif
 };

 void mmc_attach_bus(struct mmc_host *host, const struct mmc_bus_ops *ops);
@@ -59,6 +66,17 @@ void mmc_power_up(struct mmc_host *host, u32 ocr);
 void mmc_power_off(struct mmc_host *host);
 void mmc_power_cycle(struct mmc_host *host, u32 ocr);
 void mmc_set_initial_state(struct mmc_host *host);
+#ifdef CONFIG_ASCEND_HISI_MMC
+void mmc_power_up_vcc(struct mmc_host *host, u32 ocr);
+void mmc_power_off_vcc(struct mmc_host *host);
+#else
+static inline void mmc_power_up_vcc(struct mmc_host *host, u32 ocr) {}
+static inline void mmc_power_off_vcc(struct mmc_host *host) {}
+#endif
+void mmc_set_ios(struct mmc_host *host);
+void mmc_bus_get(struct mmc_host *host);
+void mmc_bus_put(struct mmc_host *host);
+

 static inline void mmc_delay(unsigned int ms)
 {
diff --git a/drivers/mmc/core/hisi_core_mmc.c b/drivers/mmc/core/hisi_core_mmc.c
new file mode 100755
index 000000000..49a94374b
--- /dev/null
+++ b/drivers/mmc/core/hisi_core_mmc.c
@@ -0,0 +1,818 @@
+// SPDX-License-Identifier: GPL-2.0
+#include <linux/device.h>
+#include <linux/delay.h>
+#include <linux/err.h>
+#include <linux/pm_runtime.h>
+
+
+#include <linux/mmc/card.h>
+#include <linux/mmc/host.h>
+#include <linux/version.h>
+#include <linux/mmc/mmc.h>
+#include "core.h"
+#include "bus.h"
+#include "host.h"
+#include "sdio_bus.h"
+
+#include "mmc_ops.h"
+#include "sd_ops.h"
+#include "sdio_ops.h"
+#include "hisi_core_mmc.h"
+
+
+void mmc_power_up_vcc(struct mmc_host *host, u32 ocr)
+{
+	if (host->ios.power_mode == MMC_POWER_UP)
+		return;
+
+	host->ios.vdd = fls(ocr) - 1;
+
+	/*
+	 * This delay should be sufficient to allow the power supply
+	 * to reach the minimum voltage.
+	 */
+	mmc_delay(10);
+
+	host->ios.clock = host->ios.clock_store;
+	host->ios.power_mode = MMC_POWER_UP;
+	mmc_set_ios(host);
+
+	/*
+	 * This delay must be at least 74 clock sizes, or 1 ms, or the
+	 * time required to reach a stable voltage.
+	 */
+	mmc_delay(10);
+
+}
+
+
+void mmc_power_off_vcc(struct mmc_host *host)
+{
+	if (host->ios.power_mode == MMC_POWER_OFF)
+		return;
+
+	/*store the ios.clock which will be used in mmc_power_up_vcc*/
+	host->ios.clock_store = host->ios.clock;
+	host->ios.clock = 0;
+	host->ios.vdd = 0;
+
+	host->ios.power_mode = MMC_POWER_OFF;
+
+	mmc_set_ios(host);
+
+	/*
+	 * Some configurations, such as the 802.11 SDIO card in the OLPC
+	 * XO-1.5, require a short delay after poweroff before the card
+	 * can be successfully turned on again.
+	 */
+	mmc_delay(1);
+}
+
+
+int mmc_card_awake(struct mmc_host *host)
+{
+	int err = -ENOSYS;
+
+	mmc_bus_get(host);
+
+	if (host->bus_ops && !host->bus_dead && host->bus_ops->awake)
+		err = host->bus_ops->awake(host);
+
+	mmc_bus_put(host);
+
+	return err;
+}
+EXPORT_SYMBOL(mmc_card_awake);
+
+int mmc_card_sleep(struct mmc_host *host)
+{
+	int err = -ENOSYS;
+
+	mmc_bus_get(host);
+
+	if (host->bus_ops && !host->bus_dead && host->bus_ops->sleep)
+		err = host->bus_ops->sleep(host);
+
+	mmc_bus_put(host);
+
+	return err;
+}
+EXPORT_SYMBOL(mmc_card_sleep);
+
+int mmc_card_can_sleep(struct mmc_host *host)
+{
+	struct mmc_card *card = host->card;
+
+	if (card && mmc_card_mmc(card) && card->ext_csd.rev >= 3)
+		return 1;
+	return 0;
+}
+EXPORT_SYMBOL(mmc_card_can_sleep);
+
+
+
+/*
+ * Turn the cache ON/OFF.
+ * Turning the cache OFF shall trigger flushing of the data
+ * to the non-volatile storage.
+ * This function should be called with host claimed
+ */
+int mmc_cache_ctrl(struct mmc_host *host, u8 enable)
+{
+	struct mmc_card *card = host->card;
+	unsigned int timeout;
+	int err = 0;
+
+	if (!(host->caps2 & MMC_CAP2_CACHE_CTRL) ||
+		      mmc_card_is_removable(host))
+		return err;
+
+	if (card && mmc_card_mmc(card) &&
+			(card->ext_csd.cache_size > 0)) {
+		enable = !!enable;
+
+		if (card->ext_csd.cache_ctrl ^ enable) {
+			timeout = enable ? card->ext_csd.generic_cmd6_time : 0;
+			err = mmc_switch(card, EXT_CSD_CMD_SET_NORMAL,
+					EXT_CSD_CACHE_CTRL, enable, timeout);
+			if (err)
+				pr_err("%s: cache %s error %d\n",
+						mmc_hostname(card->host),
+						enable ? "on" : "off",
+						err);
+			else
+				card->ext_csd.cache_ctrl = enable;
+		}
+	}
+
+	return err;
+}
+EXPORT_SYMBOL(mmc_cache_ctrl);
+
+int mmc_card_sleepawake(struct mmc_host *host, int sleep)
+{
+	struct mmc_command cmd = {0};
+	struct mmc_card *card = host->card;
+	int err;
+	unsigned long timeout;
+
+	if (sleep)
+		(void)mmc_deselect_cards(host);
+
+	cmd.opcode = MMC_SLEEP_AWAKE;
+	cmd.arg = card->rca << 16;
+	if (sleep)
+		cmd.arg |= 1 << 15;
+
+	cmd.flags = MMC_RSP_R1B | MMC_CMD_AC;
+	err = mmc_wait_for_cmd(host, &cmd, 0);
+	if (err)
+		return err;
+
+	/* we use the the check busy method to reduce sr time*/
+	if (host->ops->card_busy) {
+		timeout = jiffies +
+		msecs_to_jiffies(DIV_ROUND_UP(card->ext_csd.sa_timeout, 10000));
+		while (host->ops->card_busy(host)) {
+			/*
+			 *Timeout if the device
+			 *never leaves the program state.
+			 */
+			if (time_after(jiffies, timeout)) {
+				pr_err("%s: wait card not busy time out! %s\n",
+				mmc_hostname(host), __func__);
+				break;
+			}
+		}
+	} else {
+		/*
+		 * If the host does not wait while the card signals busy,
+		 * then we will have to wait the sleep/awake timeout.
+		 * Note, we cannot use the
+		 * SEND_STATUS command to poll the status
+		 * because that command (and most
+		 * others) is invalid while the card sleeps.
+		 */
+		if (!(host->caps & MMC_CAP_WAIT_WHILE_BUSY))
+			mmc_delay(DIV_ROUND_UP(card->ext_csd.sa_timeout,
+				10000));
+	}
+
+	if (!sleep)
+		err = mmc_select_card(card);
+
+	return err;
+}
+
+void mmc_decode_ext_csd_emmc50(struct mmc_card *card, u8 *ext_csd)
+{
+	card->ext_csd.raw_sleep_noti_time =
+		ext_csd[EXT_CSD_SLEEP_NOTIFICATION_TIME];
+	if (card->ext_csd.raw_sleep_noti_time > 0 &&
+		card->ext_csd.raw_sleep_noti_time <= 0x17) {
+		card->ext_csd.sleep_notification_time =
+		/* ms, raw_sleep_noti_time Units: 10us */
+		(1 << card->ext_csd.raw_sleep_noti_time) / 100;
+		pr_debug("%s: support SLEEP_NOTIFICATION.",
+			mmc_hostname(card->host));
+		pr_debug("sleep_notification_time=%d ms\n",
+			card->ext_csd.sleep_notification_time);
+	}
+
+#ifdef CONFIG_MMC_CQ_HCI
+	card->ext_csd.cmdq_support = ext_csd[EXT_CSD_CMDQ_SUPPORT] & 0x1;
+	if (card->ext_csd.cmdq_support) {
+		card->ext_csd.cmdq_depth =
+			(ext_csd[EXT_CSD_CMDQ_DEPTH] & 0x1F) + 1;
+		pr_info("%s: %s: CMDQ supported: depth: %d\n",
+			mmc_hostname(card->host), __func__,
+			card->ext_csd.cmdq_depth);
+	}
+#endif
+}
+
+void mmc_decode_ext_csd_emmc51(struct mmc_card *card, u8 *ext_csd)
+{
+	if (card->ext_csd.rev < 8)
+		return;
+	/* eMMC5.1*/
+	/* strobe */
+	if (ext_csd[EXT_CSD_STROBE_SUPPORT]) {
+		card->ext_csd.strobe_enhanced =
+			ext_csd[EXT_CSD_STROBE_SUPPORT];
+		if (card->ext_csd.strobe_enhanced)
+			pr_info("%s: support STROBE_ENHANCED.\n",
+				mmc_hostname(card->host));
+		else
+			pr_warn("%s: EXT_CSD_STROBE_SUPPORT bit is not set\n",
+				mmc_hostname(card->host));
+	}
+	/* cache flush barrier */
+	if (ext_csd[EXT_CSD_BARRIER_SUPPORT]) {
+		card->ext_csd.cache_flush_barrier =
+			ext_csd[EXT_CSD_BARRIER_SUPPORT];
+		if (card->ext_csd.cache_flush_barrier)
+			pr_info("%s: support BARRIER_SUPPORT.\n",
+				mmc_hostname(card->host));
+	}
+	/* cache flush policy */
+	if (ext_csd[EXT_CSD_CACHE_FLUSH_POLICY]) {
+		card->ext_csd.cache_flush_policy =
+			ext_csd[EXT_CSD_CACHE_FLUSH_POLICY] &
+			EXT_CSD_CACHE_FLUSH_POLICY_FIFO;
+		if (card->ext_csd.cache_flush_policy)
+			pr_info("%s: support CACHE_FLUSH_POLICY_FIFO.\n",
+				mmc_hostname(card->host));
+	}
+	/* bkops auto */
+	if (ext_csd[EXT_CSD_BKOPS_SUPPORT] & 0x1) {
+		card->ext_csd.bkops_auto_en =
+			ext_csd[EXT_CSD_BKOPS_EN] & EXT_CSD_BKOPS_AUTO_EN;
+		if (!card->ext_csd.bkops_auto_en)
+			pr_info("%s: BKOPS_AUTO_EN bit is not set.\n",
+				mmc_hostname(card->host));
+	}
+	/* rpmb 8k */
+	if (ext_csd[EXT_CSD_WR_REL_PARAM] & EXT_CSD_RPMB_REL_WR_EN)
+		pr_info("%s: support RPMB 8K Bytes read/write.\n",
+			mmc_hostname(card->host));
+
+	pr_info("%s: EXT_CSD revision 0x%02x pre_eol_info = 0x%02X\n",
+		mmc_hostname(card->host), card->ext_csd.rev,
+		card->ext_csd.pre_eol_info);
+	pr_info("device_life_time_est_typ_a = 0x%02X\n",
+		card->ext_csd.device_life_time_est_typ_a);
+	pr_info("device_life_time_est_typ_b = 0x%02X.\n",
+		card->ext_csd.device_life_time_est_typ_b);
+}
+
+
+/* enable feature for emmc5.0 or later contains cmdq,barrier and bkops */
+int mmc_init_card_enable_feature(struct mmc_card *card)
+{
+	int err = 0;
+	struct mmc_host *host = card->host;
+
+	/*
+	 * Enable command queue (if supported)
+	 */
+	if (card->ext_csd.cmdq_support) {
+		if (host->caps2 & MMC_CAP2_CMD_QUEUE) {
+			card->ext_csd.cmdq_mode_en = 1;
+			pr_info("%s: cmdq_mode_en=%d\n",
+				__func__, card->ext_csd.cmdq_mode_en);
+		} else {
+			card->ext_csd.cmdq_mode_en = 0;
+		}
+	}
+
+	/*
+	 * Enable Barrier feature (if supported)
+	 */
+	if (card->ext_csd.cache_flush_barrier &&
+		(host->caps2 & MMC_CAP2_CACHE_FLUSH_BARRIER)) {
+		err = mmc_switch(card, EXT_CSD_CMD_SET_NORMAL,
+				EXT_CSD_BARRIER_CTRL, 1,
+				card->ext_csd.generic_cmd6_time);
+		if (err && err != -EBADMSG)
+			return err;
+		if (err) {
+			pr_warn("%s: Enabling cache flush barrier failed\n",
+				   mmc_hostname(card->host));
+			err = 0;
+			card->ext_csd.cache_flush_barrier_en = 0;
+		} else {
+			card->ext_csd.cache_flush_barrier_en = 1;
+			pr_info("%s: cache_flush_barrier_en=%d\n",
+				__func__, card->ext_csd.cache_flush_barrier_en);
+		}
+	}
+
+	/*
+	 * Enable BKOPS AUTO  feature (if supported)
+	 */
+	if ((host->caps2 & MMC_CAP2_BKOPS_AUTO_CTRL) &&
+		(host->pm_flags & MMC_PM_KEEP_POWER) &&
+		card->ext_csd.bkops && (card->ext_csd.rev >= 8)) {
+#ifdef CONFIG_HISI_MMC_MANUAL_BKOPS
+		if (card->cid.manfid == CID_MANFID_MICRON) {
+			err = mmc_switch(card, EXT_CSD_CMD_SET_NORMAL,
+				EXT_CSD_BKOPS_EN, EXT_CSD_BKOPS_MANUAL_EN |
+				EXT_CSD_BKOPS_AUTO_EN,
+				card->ext_csd.generic_cmd6_time);
+		} else {
+#endif
+			err = mmc_switch(card, EXT_CSD_CMD_SET_NORMAL,
+					EXT_CSD_BKOPS_EN, EXT_CSD_BKOPS_AUTO_EN,
+					card->ext_csd.generic_cmd6_time);
+#ifdef CONFIG_HISI_MMC_MANUAL_BKOPS
+		}
+#endif
+		if (err && err != -EBADMSG)
+			return err;
+		if (err) {
+			pr_warn("%s: BKOPS_AUTO_EN failed\n",
+				   mmc_hostname(card->host));
+			err = 0;
+			card->ext_csd.bkops_auto_en = 0;
+#ifdef CONFIG_HISI_MMC_MANUAL_BKOPS
+			card->ext_csd.man_bkops_en = 0;
+#endif
+		} else {
+			card->ext_csd.bkops_auto_en = 1;
+#ifdef CONFIG_HISI_MMC_MANUAL_BKOPS
+			if (card->cid.manfid == CID_MANFID_MICRON)
+				card->ext_csd.man_bkops_en = 1;
+#endif
+			pr_info("%s: support BKOPS_AUTO_EN, bkops_auto_en=%d\n",
+			__func__, card->ext_csd.bkops_auto_en);
+		}
+	}
+
+	return err;
+}
+
+/* suspend operation on hisi platform, conclude disable cmdq
+ * swich to normal partion, stop bkops and so on
+ */
+int mmc_suspend_hisi_operation(struct mmc_host *host)
+{
+	unsigned long timeout = 8000;
+	u8 part_config;
+	int err = 0;
+
+	if (mmc_card_cmdq(host->card)) {
+		/* wait for cmdq req handle done.*/
+		while (host->cmdq_ctx.active_reqs) {
+			if (timeout == 0) {
+				pr_err("%s: wait reqs timeout !\n", __func__);
+				err = -ETIMEDOUT;
+				goto err_handle;
+			}
+			timeout--;
+			mdelay(1);
+		}
+
+		if (host->cmdq_ops->disable) {
+			err = host->cmdq_ops->disable(host, true);
+			if (err) {
+				pr_err("%s: cmdq disable failed.\n", __func__);
+				goto err_handle;
+			}
+		}
+
+		err = mmc_switch(host->card, EXT_CSD_CMD_SET_NORMAL,
+			EXT_CSD_CMDQ_MODE, 0,
+			host->card->ext_csd.generic_cmd6_time);
+
+		if (err) {
+			pr_err("%s: disable device cmdq failed.\n", __func__);
+			pr_err("active_reqs is %lu,cmdq status is %d.\n",
+				host->cmdq_ctx.active_reqs,
+				mmc_card_cmdq(host->card));
+			/* we re-enable the device cmdq feature just in case;
+			 * For example:
+			 * 1.rpmb_access disable the device cmdq feature;
+			 * 2.disable device cmdq feqture failed
+			 *   here in _mmc_suspend;
+			 * 3.if we don't re-enable device cmdq feature,
+			 *   next normal req
+			 * will fail;
+			 */
+			(void)mmc_switch(host->card, EXT_CSD_CMD_SET_NORMAL,
+				EXT_CSD_CMDQ_MODE, 1,
+				host->card->ext_csd.generic_cmd6_time);
+
+			if (host->cmdq_ops->enable) {
+				if (host->cmdq_ops->enable(host))
+					pr_err("%s %d: cmdq enable failed.\n",
+						__func__, __LINE__);
+			}
+			goto err_handle;
+		}
+
+		mmc_card_clr_cmdq(host->card);
+	}
+
+	/*
+	 * Ensure eMMC user default partition is enabled,because cmd5 will be
+	 * a illegal cmd when device is in rpmb partition.
+	 */
+	if (host->card->ext_csd.part_config & EXT_CSD_PART_CONFIG_ACC_MASK) {
+		pr_info("%s need to swich to default partition, current part_config is %d\n",
+			__func__, host->card->ext_csd.part_config);
+		part_config = host->card->ext_csd.part_config;
+		part_config &= ~EXT_CSD_PART_CONFIG_ACC_MASK;
+		err = mmc_switch(host->card, EXT_CSD_CMD_SET_NORMAL,
+				EXT_CSD_PART_CONFIG,
+				part_config,
+				host->card->ext_csd.part_time);
+		if (err) {
+			pr_err("%s swich to default partition failed, err is %d\n",
+				__func__, err);
+			goto err_handle;
+		}
+		host->card->ext_csd.part_config = part_config;
+	}
+
+	if (mmc_card_doing_bkops(host->card) && mmc_stop_bkops(host->card))
+		err = -1;
+err_handle:
+	return err;
+}
+#pragma GCC diagnostic push
+#pragma GCC diagnostic ignored "-Wunused-variable"
+int hisi_mmc_reset(struct mmc_host *host)
+{
+	int ret;
+	struct mmc_card *card = host->card;
+
+	pr_info("%s enter\n", __func__);
+
+	if (!host->ops->hw_reset)
+		return -EOPNOTSUPP;
+
+	host->ops->hw_reset(host);
+
+	mmc_power_off(host);
+	mdelay(200);
+	mmc_power_up(host, card->ocr);
+
+	if (host->bus_ops->power_restore == NULL)
+		/* adapt miniV2, lp API in future */
+		ret = host->bus_ops->reinit_card(host);
+	else
+		ret = host->bus_ops->power_restore(host);
+	pr_info("%s exit,ret=%d\n", __func__, ret);
+	return ret;
+}
+#pragma GCC diagnostic pop
+
+/**
+ * __mmc_send_status_direct - send the cmd 13 to device and get the device
+ * status;it can be called when the irq system cannot work.
+ */
+static int __mmc_send_status_direct(struct mmc_card *card, u32 *status,
+				    bool ignore_crc)
+{
+	int err;
+	struct mmc_command cmd = {0};
+	struct mmc_request mrq = {NULL};
+	struct mmc_host *host;
+
+	WARN_ON(!card);
+	WARN_ON(!card->host);
+
+	host = card->host;
+	cmd.opcode = MMC_SEND_STATUS;
+	if (!mmc_host_is_spi(card->host))
+		cmd.arg = card->rca << 16;
+	cmd.flags = MMC_RSP_SPI_R2 | MMC_RSP_R1 | MMC_CMD_AC;
+	if (ignore_crc)
+		cmd.flags &= ~MMC_RSP_CRC;
+
+	cmd.data = NULL;
+	cmd.busy_timeout = MMC_TIMEOUT_INVALID;
+	mrq.cmd = &cmd;
+	memset(cmd.resp, 0, sizeof(cmd.resp));
+	cmd.retries = 0;
+	mrq.cmd->error = 0;
+	mrq.cmd->mrq = &mrq;
+
+	if (host->ops->send_cmd_direct) {
+		err = host->ops->send_cmd_direct(host, &mrq);
+		if (err)
+			return err;
+	}
+
+	if (status)
+		*status = cmd.resp[0];
+
+	return 0;
+}
+
+/**
+ * mmc_switch_irq_safe - mmc_switch func for hisi platform;
+ * it can be called when the irq system cannot work.
+ */
+int mmc_switch_irq_safe(struct mmc_card *card, u8 set, u8 index, u8 value)
+{
+	struct mmc_host *host;
+	int err;
+	struct mmc_command cmd = {0};
+	struct mmc_request mrq = {NULL};
+	u32 status = 0;
+	u32 cmd_retries = 10;
+
+	WARN_ON(!card);
+	WARN_ON(!card->host);
+	host = card->host;
+
+	cmd.opcode = MMC_SWITCH;
+	cmd.arg = (MMC_SWITCH_MODE_WRITE_BYTE << 24) |
+		  (index << 16) |
+		  (value << 8) |
+		  set;
+	cmd.flags = MMC_CMD_AC;
+	cmd.flags |= MMC_RSP_SPI_R1 | MMC_RSP_R1;
+
+	if (index == EXT_CSD_SANITIZE_START)
+		cmd.sanitize_busy = true;
+
+	cmd.data = NULL;
+	cmd.busy_timeout = MMC_TIMEOUT_INVALID;
+	mrq.cmd = &cmd;
+	memset(cmd.resp, 0, sizeof(cmd.resp));
+	cmd.retries = 0;
+	mrq.cmd->error = 0;
+	mrq.cmd->mrq = &mrq;
+
+	if (host->ops->send_cmd_direct) {
+		/* this func will check busy after data send */
+		err = host->ops->send_cmd_direct(host, &mrq);
+		if (err)
+			return err;
+	}
+	do {
+		err = __mmc_send_status_direct(card, &status, (bool)false);
+		if (err)
+			return err;
+
+		/* only check the response status here so we olny send
+		 * cmd13 10 times
+		 */
+		if (--cmd_retries == 0) {
+			pr_err("%s: Card stuck in programming state! %s\n",
+				mmc_hostname(host), __func__);
+			return -ETIMEDOUT;
+		}
+	} while (R1_CURRENT_STATE(status) == R1_STATE_PRG);
+
+	if (status & 0xFDFFA000)
+		pr_warn("%s: unexpected status %#x after switch\n",
+			mmc_hostname(host), status);
+	if (status & R1_SWITCH_ERROR)
+		return -EBADMSG;
+
+	return 0;
+}
+
+/**
+ *	mmc_try_claim_host - try exclusively to claim a host
+ *	@host: mmc host to claim
+ *
+ *	Returns %1 if the host is claimed, %0 otherwise.
+ */
+int mmc_try_claim_host(struct mmc_host *host)
+{
+	int claimed_host = 0;
+	unsigned long flags;
+	bool pm = false;
+
+	if (!spin_trylock_irqsave(&host->lock, flags))
+		return 0;
+
+	if (!host->claimed || host->claimer == (struct mmc_ctx *)current) {
+		host->claimed = 1;
+		host->claimer->task = current;
+		host->claim_cnt += 1;
+		claimed_host = 1;
+		if (host->claim_cnt == 1)
+			pm = true;
+	}
+	spin_unlock_irqrestore(&host->lock, flags);
+
+	if (pm)
+		pm_runtime_get_sync(mmc_dev(host));
+
+	return claimed_host;
+}
+EXPORT_SYMBOL(mmc_try_claim_host);
+/*
+ * This is a helper function, which fetches a runtime pm reference for the
+ * card device and also claims the host.
+ */
+int mmc_get_card_hisi(struct mmc_card *card, bool use_irq)
+{
+	int claimed = 0;
+	unsigned int try_count = 200000;
+	int ret = 0;
+
+	do {
+		claimed = mmc_try_claim_host(card->host);
+		if (claimed)
+			break;
+		udelay(10);
+	} while (--try_count > 0);
+	if (!claimed) {
+		pr_err("%s try to claim host failed\n", __func__);
+		ret = -EIO;
+		goto out;
+	}
+
+	/* mmc has suspended,no more operation */
+	if (mmc_card_suspended(card)) {
+		pr_err("%s mmc has suspended;\n", __func__);
+		ret = -EHOSTDOWN;
+		goto release_host;
+	}
+
+#ifdef CONFIG_MMC_CQ_HCI
+	if (card->ext_csd.cmdq_mode_en && cmdq_is_reset(card->host)) {
+		pr_err("%s cmdq is in reset process\n", __func__);
+		ret = -EHOSTUNREACH;
+		goto release_host;
+	}
+	if (use_irq)
+		ret = mmc_blk_cmdq_hangup(card);
+	else
+		ret = mmc_blk_cmdq_halt(card);
+
+	if (ret) {
+		pr_err("%s: cmdq hangup|halt err.\n", __func__);
+		ret = -EIO;
+		goto release_host;
+	}
+
+	if (mmc_card_doing_bkops(card) && mmc_stop_bkops(card))
+		pr_err("%s: mmc_stop_bkops failed!\n", __func__);
+#endif /* CONFIG_MMC_CQ_HCI */
+	return ret;
+release_host:
+	mmc_release_host(card->host);
+out:
+	return ret;
+}
+
+void mmc_put_card_irq_safe(struct mmc_card *card)
+{
+#ifdef CONFIG_MMC_CQ_HCI
+	mmc_blk_cmdq_dishalt(card);
+#endif
+	mmc_release_host(card->host);
+}
+
+
+/*
+ * Flush the cache to the non-volatile storage.
+ */
+int mmc_flush_cache_direct(struct mmc_card *card)
+{
+	struct mmc_host *host = card->host;
+	int err = 0;
+
+	if (!(host->caps2 & MMC_CAP2_CACHE_CTRL))
+		return err;
+
+	if (mmc_card_mmc(card) &&
+			(card->ext_csd.cache_size > 0) &&
+			(card->ext_csd.cache_ctrl & 1)) {
+
+		err = mmc_switch_irq_safe(card, EXT_CSD_CMD_SET_NORMAL,
+			EXT_CSD_FLUSH_CACHE, 1);
+		if (err)
+			pr_err("%s: cache flush error %d\n",
+					mmc_hostname(card->host), err);
+		else
+			pr_info("%s success!\n", __func__);
+	}
+
+	return err;
+}
+
+
+/*********************sd ops begin**********************/
+static int mmc_do_sd_reset(struct mmc_host *host)
+{
+	struct mmc_card *card = host->card;
+
+	if (!host->bus_ops->power_restore)
+		return -EOPNOTSUPP;
+
+	if (!card)
+		return -EINVAL;
+
+	/* hw_reset for ip reset */
+	if (host->ops->hw_reset)
+		host->ops->hw_reset(host);
+
+	/* Only for K930/920 SD slow down clk*/
+	if (host->ops->slowdown_clk)
+		host->ops->slowdown_clk(host, host->ios.timing);
+
+	mmc_power_off(host);
+	mmc_set_clock(host, host->f_init);
+	/* Wait at least 200 ms */
+	mmc_delay(200);
+	mmc_power_up(host, host->card->ocr);
+	(void)mmc_select_voltage(host, host->card->ocr);
+
+	return host->bus_ops->power_restore(host);
+}
+
+int mmc_sd_reset(struct mmc_host *host)
+{
+	return mmc_do_sd_reset(host);
+}
+EXPORT_SYMBOL(mmc_sd_reset);
+
+/* low speed card, set frequency to 25M */
+void mmc_select_new_sd(struct mmc_card *card)
+{
+	unsigned int max_dtr;
+
+	mmc_set_timing(card->host, MMC_TIMING_NEW_SD);
+	max_dtr = (unsigned int)-1;
+	if (max_dtr > card->csd.max_dtr)
+		max_dtr = card->csd.max_dtr;
+	mmc_set_clock(card->host, max_dtr);
+}
+
+/*********************wifi ops begin**********************/
+int mmc_power_save_host_for_wifi(struct mmc_host *host)
+{
+	int ret = 0;
+
+#ifdef CONFIG_MMC_DEBUG
+	pr_info("%s: %s: powering down\n", mmc_hostname(host), __func__);
+#endif
+
+	mmc_bus_get(host);
+
+	if ((!host->ops || !host->ops->set_ios) ||
+	(host->bus_dead)) {
+		mmc_bus_put(host);
+		return -EINVAL;
+	}
+
+	mmc_bus_put(host);
+
+	mmc_power_off(host);
+
+	return ret;
+}
+EXPORT_SYMBOL(mmc_power_save_host_for_wifi);
+
+int mmc_power_restore_host_for_wifi(struct mmc_host *host)
+{
+	int ret = 0;
+
+#ifdef CONFIG_MMC_DEBUG
+	pr_info("%s: %s: powering up\n", mmc_hostname(host), __func__);
+#endif
+
+	mmc_bus_get(host);
+
+	if ((!host->ops || !host->ops->set_ios) ||
+	 (host->bus_dead)) {
+		mmc_bus_put(host);
+		return -EINVAL;
+	}
+
+	mmc_power_up(host, host->card->ocr);
+
+	mmc_bus_put(host);
+
+	return ret;
+}
+EXPORT_SYMBOL(mmc_power_restore_host_for_wifi);
diff --git a/drivers/mmc/core/hisi_core_mmc.h b/drivers/mmc/core/hisi_core_mmc.h
new file mode 100755
index 000000000..f662e534b
--- /dev/null
+++ b/drivers/mmc/core/hisi_core_mmc.h
@@ -0,0 +1,52 @@
+/* SPDX-License-Identifier: GPL-2.0 */
+#ifndef _HISI_CORE_MMC_H
+#define _HISI_CORE_MMC_H
+
+struct mmc_card;
+
+#ifdef CONFIG_ASCEND_HISI_MMC
+
+#ifdef CONFIG_MMC_CQ_HCI
+extern void mmc_blk_cmdq_dishalt(struct mmc_card *card);
+extern int cmdq_is_reset(struct mmc_host *host);
+extern int mmc_blk_cmdq_halt(struct mmc_card *card);
+#endif
+
+void mmc_decode_ext_csd_emmc50(struct mmc_card *card, u8 *ext_csd);
+void mmc_decode_ext_csd_emmc51(struct mmc_card *card, u8 *ext_csd);
+
+void mmc_select_new_sd(struct mmc_card *card);
+int mmc_init_card_enable_feature(struct mmc_card *card);
+int mmc_suspend_hisi_operation(struct mmc_host *host);
+int mmc_card_sleepawake(struct mmc_host *host, int sleep);
+int mmc_sd_reset(struct mmc_host *host);
+
+int hisi_mmc_reset(struct mmc_host *host);
+#else
+static inline void mmc_decode_ext_csd_emmc51(struct mmc_card *card, u8 *ext_csd)
+{
+
+}
+static inline void mmc_decode_ext_csd_emmc50(struct mmc_card *card, u8 *ext_csd)
+{
+
+}
+static inline void mmc_select_new_sd(struct mmc_card *card) {}
+static inline int mmc_init_card_enable_feature(struct mmc_card *card)
+{
+	return 0;
+}
+static inline int mmc_suspend_hisi_operation(struct mmc_host *host)
+{
+	return 0;
+}
+static inline int mmc_card_sleepawake(struct mmc_host *host, int sleep)
+{
+	return 0;
+}
+static inline int mmc_sd_reset(struct mmc_host *host)
+{
+	return 0;
+}
+#endif /* CONFIG_ASCEND_HISI_MMC */
+#endif /* _HISI_CORE_MMC_H */
diff --git a/drivers/mmc/core/host.c b/drivers/mmc/core/host.c
old mode 100644
new mode 100755
index f57f5de54..b29ee31e7
--- a/drivers/mmc/core/host.c
+++ b/drivers/mmc/core/host.c
@@ -348,6 +348,11 @@ int mmc_of_parse(struct mmc_host *host)

 EXPORT_SYMBOL(mmc_of_parse);

+static inline int mmc_is_ascend_hi_mci_1(struct device *dev)
+{
+	return !strncmp(dev_name(dev), "hi_mci.1", strlen("hi_mci.1"));
+}
+
 /**
  *	mmc_alloc_host - initialise the per-host structure.
  *	@extra: sizeof private data structure
@@ -374,7 +379,10 @@ struct mmc_host *mmc_alloc_host(int extra, struct device *dev)
 	}

 	host->index = err;
-
+	if (mmc_is_ascend_customized(dev)) {
+		if (mmc_is_ascend_hi_mci_1(dev))
+			host->index = 1;
+	}
 	dev_set_name(&host->class_dev, "mmc%d", host->index);

 	host->parent = dev;
@@ -383,12 +391,11 @@ struct mmc_host *mmc_alloc_host(int extra, struct device *dev)
 	device_initialize(&host->class_dev);
 	device_enable_async_suspend(&host->class_dev);

-	if (mmc_gpio_alloc(host)) {
-		put_device(&host->class_dev);
-		ida_simple_remove(&mmc_host_ida, host->index);
-		kfree(host);
-		return NULL;
-	}
+	if (!mmc_is_ascend_customized(host->parent))
+		if (mmc_gpio_alloc(host)) {
+			put_device(&host->class_dev);
+			return NULL;
+		}

 	spin_lock_init(&host->lock);
 	init_waitqueue_head(&host->wq);
@@ -441,7 +448,9 @@ int mmc_add_host(struct mmc_host *host)
 #endif

 	mmc_start_host(host);
-	mmc_register_pm_notifier(host);
+	if (!mmc_is_ascend_customized(host->parent) ||
+		!(host->pm_flags & MMC_PM_IGNORE_PM_NOTIFY))
+		mmc_register_pm_notifier(host);

 	return 0;
 }
@@ -458,7 +467,9 @@ EXPORT_SYMBOL(mmc_add_host);
  */
 void mmc_remove_host(struct mmc_host *host)
 {
-	mmc_unregister_pm_notifier(host);
+	if (!mmc_is_ascend_customized(host->parent) ||
+		!(host->pm_flags & MMC_PM_IGNORE_PM_NOTIFY))
+		mmc_unregister_pm_notifier(host);
 	mmc_stop_host(host);

 #ifdef CONFIG_DEBUG_FS
@@ -485,3 +496,19 @@ void mmc_free_host(struct mmc_host *host)
 }

 EXPORT_SYMBOL(mmc_free_host);
+
+
+int mmc_is_ascend_customized(struct device *dev)
+{
+#ifdef CONFIG_ASCEND_HISI_MMC
+	static int is_ascend_customized = -1;
+
+	if (is_ascend_customized == -1)
+		is_ascend_customized = ((dev == NULL) ? 0 :
+		of_find_property(dev->of_node, "customized", NULL) != NULL);
+	return is_ascend_customized;
+#else
+	return 0;
+#endif
+}
+EXPORT_SYMBOL(mmc_is_ascend_customized);
diff --git a/drivers/mmc/core/mmc.c b/drivers/mmc/core/mmc.c
old mode 100644
new mode 100755
index f1fe446ee..5135e665c
--- a/drivers/mmc/core/mmc.c
+++ b/drivers/mmc/core/mmc.c
@@ -28,6 +28,7 @@
 #include "quirks.h"
 #include "sd_ops.h"
 #include "pwrseq.h"
+#include "hisi_core_mmc.h"

 #define DEFAULT_CMD6_TIMEOUT_MS	500
 #define MIN_CACHE_EN_TIMEOUT_MS 1600
@@ -415,7 +416,9 @@ static int mmc_decode_ext_csd(struct mmc_card *card, u8 *ext_csd)
 	card->ext_csd.strobe_support = ext_csd[EXT_CSD_STROBE_SUPPORT];
 	card->ext_csd.raw_card_type = ext_csd[EXT_CSD_CARD_TYPE];
 	mmc_select_card_type(card);
-
+	if (mmc_is_ascend_customized(card->host->parent))
+		card->ext_csd.raw_driver_strength =
+			ext_csd[EXT_CSD_DRIVER_STRENGTH];
 	card->ext_csd.raw_s_a_timeout = ext_csd[EXT_CSD_S_A_TIMEOUT];
 	card->ext_csd.raw_erase_timeout_mult =
 		ext_csd[EXT_CSD_ERASE_TIMEOUT_MULT];
@@ -424,7 +427,9 @@ static int mmc_decode_ext_csd(struct mmc_card *card, u8 *ext_csd)
 	if (card->ext_csd.rev >= 3) {
 		u8 sa_shift = ext_csd[EXT_CSD_S_A_TIMEOUT];
 		card->ext_csd.part_config = ext_csd[EXT_CSD_PART_CONFIG];
-
+		if (mmc_is_ascend_customized(card->host->parent) &&
+			card->cid.manfid == CID_MANFID_HYNIX)
+			ext_csd[EXT_CSD_PART_SWITCH_TIME] = 0xA;
 		/* EXT_CSD value is in units of 10ms, but we store in ms */
 		card->ext_csd.part_time = 10 * ext_csd[EXT_CSD_PART_SWITCH_TIME];
 		/* Some eMMC set the value too low so set a minimum */
@@ -513,6 +518,9 @@ static int mmc_decode_ext_csd(struct mmc_card *card, u8 *ext_csd)
 			ext_csd[EXT_CSD_PWR_CL_200_195];
 		card->ext_csd.raw_pwr_cl_200_360 =
 			ext_csd[EXT_CSD_PWR_CL_200_360];
+		if (mmc_is_ascend_customized(card->host->parent))
+			card->ext_csd.raw_pwr_cl_ddr_200_360 =
+				ext_csd[EXT_CSD_PWR_CL_DDR_200_360];
 		card->ext_csd.raw_pwr_cl_ddr_52_195 =
 			ext_csd[EXT_CSD_PWR_CL_DDR_52_195];
 		card->ext_csd.raw_pwr_cl_ddr_52_360 =
@@ -622,6 +630,9 @@ static int mmc_decode_ext_csd(struct mmc_card *card, u8 *ext_csd)

 	/* eMMC v5 or later */
 	if (card->ext_csd.rev >= 7) {
+		if (mmc_is_ascend_customized(card->host->parent))
+			/*new feature for emmc v5.0*/
+			mmc_decode_ext_csd_emmc50(card, ext_csd);
 		memcpy(card->ext_csd.fwrev, &ext_csd[EXT_CSD_FIRMWARE_VERSION],
 		       MMC_FIRMWARE_LEN);
 		card->ext_csd.ffu_capable =
@@ -651,7 +662,21 @@ static int mmc_decode_ext_csd(struct mmc_card *card, u8 *ext_csd)
 				 mmc_hostname(card->host),
 				 card->ext_csd.cmdq_depth);
 		}
+	} else {
+		if (mmc_is_ascend_customized(card->host->parent)) {
+			/* 0x00 means Not defined,
+			 * write 0x00 in lower eMMC version
+			 */
+			card->ext_csd.pre_eol_info = 0x00;
+			card->ext_csd.device_life_time_est_typ_a = 0x00;
+			card->ext_csd.device_life_time_est_typ_b = 0x00;
+			card->ext_csd.cmdq_support = 0;
+			card->ext_csd.cmdq_depth = 0;
+		}
 	}
+	if (mmc_is_ascend_customized(card->host->parent))
+		/*new feature for emmc v5.1*/
+		mmc_decode_ext_csd_emmc51(card, ext_csd);
 out:
 	return err;
 }
@@ -760,6 +785,12 @@ static int mmc_compare_ext_csds(struct mmc_card *card, unsigned bus_width)
 		(card->ext_csd.raw_pwr_cl_ddr_200_360 ==
 			bw_ext_csd[EXT_CSD_PWR_CL_DDR_200_360]));

+	if (!err && mmc_is_ascend_customized(card->host->parent))
+		err =
+		!((card->ext_csd.raw_driver_strength ==
+			bw_ext_csd[EXT_CSD_DRIVER_STRENGTH]) &&
+		(card->ext_csd.raw_pwr_cl_ddr_200_360 ==
+			bw_ext_csd[EXT_CSD_PWR_CL_DDR_200_360]));
 	if (err)
 		err = -EINVAL;

@@ -794,6 +825,12 @@ MMC_DEV_ATTR(rel_sectors, "%#x\n", card->ext_csd.rel_sectors);
 MMC_DEV_ATTR(ocr, "0x%08x\n", card->ocr);
 MMC_DEV_ATTR(rca, "0x%04x\n", card->rca);
 MMC_DEV_ATTR(cmdq_en, "%d\n", card->ext_csd.cmdq_en);
+#ifdef CONFIG_ASCEND_HISI_MMC
+MMC_DEV_ATTR(life_time_est_typ_a, "0x%02X\n",
+	card->ext_csd.device_life_time_est_typ_a);
+MMC_DEV_ATTR(life_time_est_typ_b, "0x%02X\n",
+	card->ext_csd.device_life_time_est_typ_b);
+#endif

 static ssize_t mmc_fwrev_show(struct device *dev,
 			      struct device_attribute *attr,
@@ -852,6 +889,10 @@ static struct attribute *mmc_std_attrs[] = {
 	&dev_attr_rca.attr,
 	&dev_attr_dsr.attr,
 	&dev_attr_cmdq_en.attr,
+#ifdef CONFIG_ASCEND_HISI_MMC
+	&dev_attr_life_time_est_typ_a.attr,
+	&dev_attr_life_time_est_typ_b.attr,
+#endif
 	NULL,
 };
 ATTRIBUTE_GROUPS(mmc_std);
@@ -1490,6 +1531,65 @@ err:
 	return err;
 }

+#ifdef CONFIG_ASCEND_HISI_MMC
+static int hisi_mmc_select_hs200(struct mmc_card *card)
+{
+	struct mmc_host *host = card->host;
+	unsigned int old_timing;
+	int err = -EINVAL;
+	u8 val;
+
+	mmc_select_driver_type(card);
+
+	/*
+	 * Set the bus width(4 or 8) with host's support and
+	 * switch to HS200 mode if bus width is set successfully.
+	 */
+	err = mmc_select_bus_width(card);
+	if (err > 0) {
+		val = EXT_CSD_TIMING_HS200 |
+		      card->drive_strength << EXT_CSD_DRV_STR_SHIFT;
+		err = __mmc_switch(card, EXT_CSD_CMD_SET_NORMAL,
+				   EXT_CSD_HS_TIMING, val,
+				   card->ext_csd.generic_cmd6_time, 0,
+				   true, false, true);
+		if (err)
+			goto err;
+		old_timing = host->ios.timing;
+		mmc_set_timing(host, MMC_TIMING_MMC_HS200);
+		/*need to change voltage after set timing on hisi plat*/
+		if (card->mmc_avail_type & EXT_CSD_CARD_TYPE_HS200_1_2V)
+			err = mmc_set_signal_voltage(host,
+				MMC_SIGNAL_VOLTAGE_120);
+		if (card->mmc_avail_type & EXT_CSD_CARD_TYPE_HS200_1_8V)
+			err = mmc_set_signal_voltage(host,
+				MMC_SIGNAL_VOLTAGE_180);
+		/* If fails try again during next card power cycle */
+		if (err)
+			goto err;
+	}
+err:
+	if (err)
+		pr_err("%s: %s failed, error %d\n", mmc_hostname(card->host),
+		       __func__, err);
+	return err;
+}
+
+static inline int mmc_select_hs200_adapt(struct mmc_card *card)
+{
+	if (mmc_is_ascend_customized(card->host->parent))
+		return hisi_mmc_select_hs200(card);
+	return mmc_select_hs200(card);
+}
+
+#else
+
+static inline int mmc_select_hs200_adapt(struct mmc_card *card)
+{
+	return mmc_select_hs200(card);
+}
+#endif /* CONFIG_ASCEND_HISI_MMC */
+
 /*
  * Activate High Speed, HS200 or HS400ES mode if supported.
  */
@@ -1503,10 +1603,12 @@ static int mmc_select_timing(struct mmc_card *card)
 	if (card->mmc_avail_type & EXT_CSD_CARD_TYPE_HS400ES)
 		err = mmc_select_hs400es(card);
 	else if (card->mmc_avail_type & EXT_CSD_CARD_TYPE_HS200)
-		err = mmc_select_hs200(card);
+		err = mmc_select_hs200_adapt(card);
 	else if (card->mmc_avail_type & EXT_CSD_CARD_TYPE_HS)
 		err = mmc_select_hs(card);
-
+	else if (mmc_is_ascend_customized(card->host->parent) &&
+		card->mmc_avail_type == 0)
+		mmc_select_new_sd(card);
 	if (err && err != -EBADMSG)
 		return err;

@@ -1549,6 +1651,7 @@ static int mmc_init_card(struct mmc_host *host, u32 ocr,
 	struct mmc_card *oldcard)
 {
 	struct mmc_card *card;
+	int cache_size_is_valid;
 	int err;
 	u32 cid[4];
 	u32 rocr;
@@ -1740,6 +1843,21 @@ static int mmc_init_card(struct mmc_host *host, u32 ocr,
 			card->ext_csd.power_off_notification = EXT_CSD_POWER_ON;
 	}

+#ifdef CONFIG_ASCEND_HISI_MMC
+	/*
+	 * Enable Enhanced strobe
+	 */
+	if (mmc_is_ascend_customized(card->host->parent) &&
+		card->ext_csd.strobe_enhanced) {
+		if (host->caps2 & MMC_CAP2_ENHANCED_STROBE) {
+			card->ext_csd.strobe_enhanced_en = 1;
+			pr_err("%s: strobe_enhanced_en=%d\n", __func__,
+			       card->ext_csd.strobe_enhanced_en);
+		} else {
+			card->ext_csd.strobe_enhanced_en = 0;
+		}
+	}
+#endif
 	/*
 	 * Select timing interface
 	 */
@@ -1763,6 +1881,10 @@ static int mmc_init_card(struct mmc_host *host, u32 ocr,
 			if (err)
 				goto free_card;
 		}
+	} else {
+		if (mmc_is_ascend_customized(card->host->parent))
+			/*to support mmc card*/
+			mmc_select_bus_width(card);
 	}

 	/*
@@ -1796,7 +1918,15 @@ static int mmc_init_card(struct mmc_host *host, u32 ocr,
 	 * sudden power failure tests. Let's extend the timeout to a minimum of
 	 * DEFAULT_CACHE_EN_TIMEOUT_MS and do it for all cards.
 	 */
-	if (card->ext_csd.cache_size > 0) {
+#ifdef CONFIG_ASCEND_HISI_MMC
+	if (mmc_is_ascend_customized(host->parent))
+		cache_size_is_valid = (host->caps2 & MMC_CAP2_CACHE_CTRL) &&
+			(card->ext_csd.cache_size > 0);
+	else
+#endif
+		cache_size_is_valid = (card->ext_csd.cache_size > 0);
+
+	if (cache_size_is_valid) {
 		unsigned int timeout_ms = MIN_CACHE_EN_TIMEOUT_MS;

 		timeout_ms = max(card->ext_csd.generic_cmd6_time, timeout_ms);
@@ -1862,6 +1992,11 @@ static int mmc_init_card(struct mmc_host *host, u32 ocr,
 		goto free_card;
 	}

+	if (mmc_is_ascend_customized(host->parent)) {
+		err = mmc_init_card_enable_feature(card);
+		if (err)
+			goto free_card;
+	}
 	if (!oldcard)
 		host->card = card;

@@ -1879,6 +2014,58 @@ static int mmc_can_sleep(struct mmc_card *card)
 	return (card && card->ext_csd.rev >= 3);
 }

+#ifdef CONFIG_ASCEND_HISI_MMC
+/*
+ * reinit card after reset in 2019.5.28
+ */
+static int mmc_reinit_card(struct mmc_host *host)
+{
+	int err;
+	struct mmc_card *card = host->card;
+
+	pr_info("enter reinit card\n");
+	mmc_flush_cache(host->card);
+
+	err = mmc_init_card(host, card->ocr, card);
+	pr_info("exit reinit card ret %d\n", err);
+	return err;
+}
+
+static int mmc_can_sleep_notify(const struct mmc_card *card)
+{
+	return card && mmc_card_mmc(card) && card->ext_csd.raw_sleep_noti_time;
+}
+static int hisi_mmc_sleep(struct mmc_host *host)
+{
+	struct mmc_card *card = host->card;
+	int err = -ENOSYS;
+
+	if (card && card->ext_csd.rev >= 3) {
+		err = mmc_card_sleepawake(host, 1);
+		if (err < 0)
+			pr_debug("%s: Error %d while putting card into sleep",
+				mmc_hostname(host), err);
+	}
+
+	return err;
+}
+
+static int mmc_awake(struct mmc_host *host)
+{
+	struct mmc_card *card = host->card;
+	int err = -ENOSYS;
+
+	if (card && card->ext_csd.rev >= 3) {
+		err = mmc_card_sleepawake(host, 0);
+		if (err < 0)
+			pr_debug("%s: Error %d while awaking sleeping card",
+		mmc_hostname(host), err);
+	}
+
+	return err;
+}
+#endif
+
 static int mmc_sleep(struct mmc_host *host)
 {
 	struct mmc_command cmd = {};
@@ -1943,7 +2130,11 @@ static int mmc_poweroff_notify(struct mmc_card *card, unsigned int notify_type)
 	/* Use EXT_CSD_POWER_OFF_SHORT as default notification type. */
 	if (notify_type == EXT_CSD_POWER_OFF_LONG)
 		timeout = card->ext_csd.power_off_longtime;
-
+#ifdef CONFIG_ASCEND_HISI_MMC
+	else if (mmc_is_ascend_customized(card->host->parent) &&
+		notify_type == EXT_CSD_SLEEP_NOTIFICATION)
+		timeout = card->ext_csd.sleep_notification_time;
+#endif
 	err = __mmc_switch(card, EXT_CSD_CMD_SET_NORMAL,
 			EXT_CSD_POWER_OFF_NOTIFICATION,
 			notify_type, timeout, 0, true, false, false);
@@ -2000,6 +2191,85 @@ static void mmc_detect(struct mmc_host *host)
 	}
 }

+#ifdef CONFIG_ASCEND_HISI_MMC
+/*suspend operation on hisi platform, conclude disable cmdq
+ *swich to normal partion and so on
+ */
+extern int mmc_suspend_hisi_operation(struct mmc_host *host);
+static int _hisi_mmc_suspend(struct mmc_host *host, bool is_suspend)
+{
+	int err = 0;
+	unsigned int notify_type = is_suspend ? EXT_CSD_SLEEP_NOTIFICATION :
+					EXT_CSD_POWER_OFF_LONG;
+
+	WARN_ON(!host);
+	WARN_ON(!host->card);
+
+	mmc_claim_host(host);
+
+	if (mmc_card_suspended(host->card))
+		goto out;
+
+	err = mmc_suspend_hisi_operation(host);
+	if (err) {
+		pr_err("%s: mmc suspend hisi operation failed.\n", __func__);
+		goto out;
+	}
+
+	err = mmc_cache_ctrl(host, 0);
+	if (err) {
+		pr_err("%s: cache disab failed.\n", __func__);
+		goto out;
+	}
+	if (mmc_can_sleep(host->card)) {
+		if (mmc_can_poweroff_notify(host->card) &&
+			((host->caps2 & MMC_CAP2_FULL_PWR_CYCLE &&
+			mmc_can_sleep_notify(host->card)) || !is_suspend)) {
+			err = mmc_poweroff_notify(host->card, notify_type);
+			if (err)
+				goto out;
+		}
+		err = mmc_card_sleep(host);
+		/*if sleep(send cmd5 failed),we reinit the card
+		 *we don't need to reinit when shut down
+		 */
+		if (err && is_suspend) {
+			pr_err("%s suspend failed err=%d\n", __func__, err);
+			mmc_power_off(host);
+			mdelay(200);
+			mmc_power_up(host, host->card->ocr);
+			(void)mmc_init_card(host, host->card->ocr, host->card);
+		}
+	} else if (!mmc_host_is_spi(host)) {
+		if (mmc_can_poweroff_notify(host->card))
+			err = mmc_poweroff_notify(host->card,
+				EXT_CSD_POWER_OFF_SHORT);
+		else
+			err = mmc_deselect_cards(host);
+	}
+
+	if (!err) {
+		if (!mmc_card_keep_power(host))
+			mmc_power_off_vcc(host);
+		mmc_card_set_suspended(host->card);
+	}
+out:
+	mmc_release_host(host);
+	return err;
+}
+#else
+static int _hisi_mmc_suspend(struct mmc_host *host, bool is_suspend)
+{
+	return 0;
+}
+#endif
+
+static bool _mmc_cache_enabled(struct mmc_host *host)
+{
+	return host->card->ext_csd.cache_size > 0 &&
+	       host->card->ext_csd.cache_ctrl & 1;
+}
+
 static int _mmc_suspend(struct mmc_host *host, bool is_suspend)
 {
 	int err = 0;
@@ -2054,6 +2324,59 @@ static int mmc_suspend(struct mmc_host *host)
 	return err;
 }

+#ifdef CONFIG_ASCEND_HISI_MMC
+static int hisi_mmc_suspend(struct mmc_host *host)
+{
+	int err;
+
+	err = _hisi_mmc_suspend(host, true);
+	if (!err) {
+		pm_runtime_disable(&host->card->dev);
+		pm_runtime_set_suspended(&host->card->dev);
+	}
+
+	return err;
+}
+
+static int hisi_enable_power_off_notify(struct mmc_host *host)
+{
+	int err = 0;
+
+	if (!mmc_card_keep_power(host))
+		mmc_power_up_vcc(host, host->card->ocr);
+
+	/* The native code enable power_off_notification
+	 * in mmc_init_card,
+	 * But we use mmc_card_awake instead of mmc_init_card
+	 * so we enable power_off_notification here.
+	 */
+	if (host->caps2 & MMC_CAP2_FULL_PWR_CYCLE
+		&& mmc_can_sleep_notify(host->card))
+		host->card->ext_csd.power_off_notification =
+			EXT_CSD_POWER_ON;
+	err = mmc_card_awake(host);/*according to K3 modify*/
+	if (err) {
+		pr_err("%s awake failed err=%d\n", __func__, err);
+		mmc_power_off(host);
+		mdelay(200);
+		mmc_power_up(host, host->card->ocr);
+		err = mmc_init_card(host, host->card->ocr, host->card);
+	} else {
+		err = mmc_cache_ctrl(host, 1);
+		if (err)
+			pr_err("%s: error %d cache on\n",
+				mmc_hostname(host), err);
+	}
+
+	return err;
+}
+#else
+static int hisi_enable_power_off_notify(struct mmc_host *host)
+{
+	return 0;
+}
+#endif
+
 /*
  * This function tries to determine if the same card is still present
  * and, if so, restore all state to it.
@@ -2066,7 +2389,8 @@ static int _mmc_resume(struct mmc_host *host)

 	if (!mmc_card_suspended(host->card))
 		goto out;
-
+	if (mmc_is_ascend_customized(host->parent))
+		err = hisi_enable_power_off_notify(host);
 	mmc_power_up(host, host->card->ocr);
 	err = mmc_init_card(host, host->card->ocr, host->card);
 	mmc_card_clr_suspended(host->card);
@@ -2091,8 +2415,12 @@ static int mmc_shutdown(struct mmc_host *host)
 		!(host->caps2 & MMC_CAP2_FULL_PWR_CYCLE))
 		err = _mmc_resume(host);

-	if (!err)
-		err = _mmc_suspend(host, false);
+	if (!err) {
+		if (mmc_is_ascend_customized(host->parent))
+			err = _hisi_mmc_suspend(host, false);
+		else
+			err = _mmc_suspend(host, false);
+	}

 	return err;
 }
@@ -2102,8 +2430,18 @@ static int mmc_shutdown(struct mmc_host *host)
  */
 static int mmc_resume(struct mmc_host *host)
 {
+	int err = 0;
+#ifdef CONFIG_ASCEND_HISI_MMC
+	if (mmc_is_ascend_customized(host->parent)) {
+		if (!(host->caps & MMC_CAP_RUNTIME_RESUME)) {
+			err = _mmc_resume(host);
+			pm_runtime_set_active(&host->card->dev);
+			pm_runtime_mark_last_busy(&host->card->dev);
+		}
+	}
+#endif
 	pm_runtime_enable(&host->card->dev);
-	return 0;
+	return err;
 }

 /*
@@ -2116,7 +2454,11 @@ static int mmc_runtime_suspend(struct mmc_host *host)
 	if (!(host->caps & MMC_CAP_AGGRESSIVE_PM))
 		return 0;

-	err = _mmc_suspend(host, true);
+	if (mmc_is_ascend_customized(host->parent))
+		err = _hisi_mmc_suspend(host, true);
+	else
+		err = _mmc_suspend(host, true);
+
 	if (err)
 		pr_err("%s: error %d doing aggressive suspend\n",
 			mmc_hostname(host), err);
@@ -2131,6 +2473,12 @@ static int mmc_runtime_resume(struct mmc_host *host)
 {
 	int err;

+#ifdef CONFIG_ASCEND_HISI_MMC
+	if (mmc_is_ascend_customized(host->parent))
+		if (!(host->caps &
+			(MMC_CAP_AGGRESSIVE_PM | MMC_CAP_RUNTIME_RESUME)))
+			return 0;
+#endif
 	err = _mmc_resume(host);
 	if (err && err != -ENOMEDIUM)
 		pr_err("%s: error %d doing runtime resume\n",
@@ -2186,6 +2534,23 @@ static const struct mmc_bus_ops mmc_ops = {
 	.hw_reset = _mmc_hw_reset,
 };

+#ifdef CONFIG_ASCEND_HISI_MMC
+static const struct mmc_bus_ops hisi_mmc_ops = {
+	.awake = mmc_awake,
+	.sleep = hisi_mmc_sleep,
+	.remove = mmc_remove,
+	.detect = mmc_detect,
+	.suspend = hisi_mmc_suspend,
+	.resume = mmc_resume,
+	.runtime_suspend = mmc_runtime_suspend,
+	.runtime_resume = mmc_runtime_resume,
+	.alive = mmc_alive,
+	.shutdown = mmc_shutdown,
+	.hw_reset = hisi_mmc_reset,
+	.reinit_card = mmc_reinit_card,
+};
+#endif
+
 /*
  * Starting point for MMC card init.
  */
@@ -2203,8 +2568,12 @@ int mmc_attach_mmc(struct mmc_host *host)
 	err = mmc_send_op_cond(host, 0, &ocr);
 	if (err)
 		return err;
-
-	mmc_attach_bus(host, &mmc_ops);
+#ifdef CONFIG_ASCEND_HISI_MMC
+	if (mmc_is_ascend_customized(host->parent))
+		mmc_attach_bus(host, &hisi_mmc_ops);
+	else
+#endif
+		mmc_attach_bus(host, &mmc_ops);
 	if (host->ocr_avail_mmc)
 		host->ocr_avail = host->ocr_avail_mmc;

diff --git a/drivers/mmc/core/sd.c b/drivers/mmc/core/sd.c
old mode 100644
new mode 100755
index 04738359e..20ca371b9
--- a/drivers/mmc/core/sd.c
+++ b/drivers/mmc/core/sd.c
@@ -28,6 +28,7 @@
 #include "mmc_ops.h"
 #include "sd.h"
 #include "sd_ops.h"
+#include "hisi_core_mmc.h"

 static const unsigned int tran_exp[] = {
 	10000,		100000,		1000000,	10000000,
@@ -164,6 +165,19 @@ static int mmc_decode_csd(struct mmc_card *card)
 		m = UNSTUFF_BITS(resp, 48, 22);
 		csd->capacity     = (1 + m) << 10;

+		/*
+		 * Bit 12 - temporary write protection
+		 * Bit 13 - permanent write protection
+		 * while do sd card detection, check write protection attribute
+		 */
+		if (mmc_is_ascend_customized(card->host->parent))
+			if (UNSTUFF_BITS(resp, 12, 1) ||
+				UNSTUFF_BITS(resp, 13, 1)) {
+				pr_info("%s, SD is on write protection\n",
+					mmc_hostname(card->host));
+				mmc_card_set_readonly(card);
+			}
+
 		csd->read_blkbits = 9;
 		csd->read_partial = 0;
 		csd->write_misalign = 0;
@@ -255,7 +269,10 @@ static int mmc_read_ssr(struct mmc_card *card)

 	for (i = 0; i < 16; i++)
 		card->raw_ssr[i] = be32_to_cpu(raw_ssr[i]);
-
+#ifdef CONFIG_ASCEND_HISI_MMC
+	if (mmc_is_ascend_customized(card->host->parent))
+		card->ssr.speed_class = UNSTUFF_BITS(raw_ssr, 440 - 384, 8);
+#endif
 	kfree(raw_ssr);

 	/*
@@ -1039,6 +1056,10 @@ retry:
 		err = mmc_sd_init_uhs_card(card);
 		if (err)
 			goto free_card;
+#ifdef CONFIG_ASCEND_HISI_MMC
+		if (mmc_is_ascend_customized(card->host->parent))
+			mmc_sd_card_set_uhs(card);
+#endif
 	} else {
 		/*
 		 * Attempt to change to high-speed (if supported)
@@ -1048,7 +1069,8 @@ retry:
 			mmc_set_timing(card->host, MMC_TIMING_SD_HS);
 		else if (err)
 			goto free_card;
-
+		else if (mmc_is_ascend_customized(card->host->parent))
+			mmc_set_timing(card->host, MMC_TIMING_NEW_SD);
 		/*
 		 * Set bus speed.
 		 */
@@ -1141,7 +1163,10 @@ static int _mmc_sd_suspend(struct mmc_host *host)
 		err = mmc_deselect_cards(host);

 	if (!err) {
-		mmc_power_off(host);
+		if (!(mmc_is_ascend_customized(host->parent)))
+			mmc_power_off(host);
+		else if (mmc_card_keep_power(host))
+			mmc_power_off(host);
 		mmc_card_set_suspended(host->card);
 	}

@@ -1178,9 +1203,23 @@ static int _mmc_sd_resume(struct mmc_host *host)

 	if (!mmc_card_suspended(host->card))
 		goto out;
+	if (mmc_is_ascend_customized(host->parent)) {
+		if (!mmc_card_keep_power(host)) {
+			mmc_power_up(host, host->card->ocr);
+			err = mmc_sd_init_card(
+				host, host->card->ocr, host->card);
+			err = mmc_sd_init_card(
+				host, host->card->ocr, host->card);
+		} else {
+			err = mmc_select_card(host->card);
+			if (err)
+				err = mmc_sd_reset(host);
+		}

-	mmc_power_up(host, host->card->ocr);
-	err = mmc_sd_init_card(host, host->card->ocr, host->card);
+	} else {
+		mmc_power_up(host, host->card->ocr);
+		err = mmc_sd_init_card(host, host->card->ocr, host->card);
+	}
 	mmc_card_clr_suspended(host->card);

 out:
@@ -1230,8 +1269,42 @@ static int mmc_sd_runtime_resume(struct mmc_host *host)
 	return 0;
 }

+#ifdef CONFIG_ASCEND_HISI_MMC
+/*********************sd ops begin**********************/
+static int mmc_do_sd_reset(struct mmc_host *host)
+{
+	struct mmc_card *card = host->card;
+
+	if (!host->bus_ops->power_restore)
+		return -EOPNOTSUPP;
+
+	if (!card)
+		return -EINVAL;
+
+	/* hw_reset for ip reset */
+	if (host->ops->hw_reset)
+		host->ops->hw_reset(host);
+
+	/* Only for K930/920 SD slow down clk*/
+	if (host->ops->slowdown_clk)
+		host->ops->slowdown_clk(host, host->ios.timing);
+
+	mmc_power_off(host);
+	mmc_set_clock(host, host->f_init);
+	/* Wait at least 200 ms */
+	mmc_delay(200);
+	mmc_power_up(host, host->card->ocr);
+	(void)mmc_select_voltage(host, host->card->ocr);
+
+	return host->bus_ops->power_restore(host);
+}
+#endif
 static int mmc_sd_hw_reset(struct mmc_host *host)
 {
+#ifdef CONFIG_ASCEND_HISI_MMC
+	if (mmc_is_ascend_customized(host->parent))
+		return mmc_do_sd_reset(host);
+#endif
 	mmc_power_cycle(host, host->card->ocr);
 	return mmc_sd_init_card(host, host->card->ocr, host->card);
 }
diff --git a/include/linux/mmc/card.h b/include/linux/mmc/card.h
old mode 100644
new mode 100755
index 8ef330027..45899066f
--- a/include/linux/mmc/card.h
+++ b/include/linux/mmc/card.h
@@ -128,6 +128,26 @@ struct mmc_ext_csd {

 	unsigned int            feature_support;
 #define MMC_DISCARD_FEATURE	BIT(0)                  /* CMD38 feature */
+#ifdef CONFIG_ASCEND_HISI_MMC
+	unsigned int		sleep_notification_time;	/* Units: ms */
+	u8			raw_vendor_feature_support;	/* 124 */
+	u8			raw_wr_rel_param;/* 166 */
+	u8			raw_rpmb_region1_size_mult;	/* 180 */
+	u8			raw_rpmb_region2_size_mult;	/* 182 */
+	u8			raw_rpmb_region3_size_mult;	/* 186 */
+	u8			raw_rpmb_region4_size_mult;	/* 188 */
+	u8			raw_sleep_noti_time;    /* 216 */
+	u8			cmdq_mode_en;			/* 15 */
+	/* 184 enhanced strobe support */
+	u8			strobe_enhanced;
+	u8			strobe_enhanced_en;
+	/* 486 barrier support */
+	u8			cache_flush_barrier;
+	u8			cache_flush_barrier_en;	/* barrier enable */
+	u8			cache_flush_policy;		/* 240 */
+	/* 163 bit(1) background op auto enable */
+	u8			bkops_auto_en;
+#endif
 };

 struct sd_scr {
@@ -145,6 +165,9 @@ struct sd_ssr {
 	unsigned int		au;			/* In sectors */
 	unsigned int		erase_timeout;		/* In milliseconds */
 	unsigned int		erase_offset;		/* In milliseconds */
+#ifdef CONFIG_ASCEND_HISI_MMC
+	unsigned int		speed_class;
+#endif
 };

 struct sd_switch_caps {
@@ -252,6 +275,18 @@ struct mmc_card {
 #define MMC_TYPE_SDIO		2		/* SDIO card */
 #define MMC_TYPE_SD_COMBO	3		/* SD combo (IO+mem) card */
 	unsigned int		state;		/* (our) card state */
+#ifdef CONFIG_ASCEND_HISI_MMC
+#define MMC_STATE_PRESENT	(1<<0)		/* present in sysfs */
+#define MMC_STATE_READONLY	(1<<1)		/* card is read-only */
+#define MMC_STATE_BLOCKADDR	(1<<2)		/* card uses block-addressing */
+#define MMC_CARD_SDXC		(1<<3)		/* card is SDXC */
+#define MMC_CARD_REMOVED	(1<<4)		/* card has been removed */
+#define MMC_STATE_DOING_BKOPS	(1<<5)		/* card is doing BKOPS */
+/* card is in ultra high speed mode */
+#define MMC_STATE_ULTRAHIGHSPEED (1<<14)
+#define MMC_STATE_SUSPENDED	(1<<6)		/* card is suspended */
+#define MMC_STATE_CMDQ         (1<<7)         /* card is in cmd queue mode */
+#endif
 	unsigned int		quirks; 	/* card quirks */
 	unsigned int		quirk_max_rate;	/* max rate set by quirks */
 #define MMC_QUIRK_LENIENT_FN0	(1<<0)		/* allow SDIO FN0 writes outside of the VS CCCR range */
@@ -259,6 +294,10 @@ struct mmc_card {
 						/* for byte mode */
 #define MMC_QUIRK_NONSTD_SDIO	(1<<2)		/* non-standard SDIO card attached */
 						/* (missing CIA registers) */
+#ifdef CONFIG_ASCEND_HISI_MMC
+/* clock gating the sdio bus will make card fail */
+#define MMC_QUIRK_BROKEN_CLK_GATING (1<<3)
+#endif
 #define MMC_QUIRK_NONSTD_FUNC_IF (1<<4)		/* SDIO card has nonstd function interfaces */
 #define MMC_QUIRK_DISABLE_CD	(1<<5)		/* disconnect CD/DAT[3] resistor */
 #define MMC_QUIRK_INAND_CMD38	(1<<6)		/* iNAND devices have broken CMD38 */
@@ -322,4 +361,70 @@ bool mmc_card_is_blockaddr(struct mmc_card *card);
 #define mmc_card_sd(c)		((c)->type == MMC_TYPE_SD)
 #define mmc_card_sdio(c)	((c)->type == MMC_TYPE_SDIO)

+#ifdef CONFIG_ASCEND_HISI_MMC
+#define MMC_CARD_CMDQ_BLK_SIZE 512
+#define EXT_CSD_PRE_EOL_INFO_NORMAL     0x01
+#define EXT_CSD_PRE_EOL_INFO_WARNING     0x02
+#define EXT_CSD_PRE_EOL_INFO_URGENT     0x03
+
+#define mmc_card_present(c)	((c)->state & MMC_STATE_PRESENT)
+#define mmc_card_readonly(c)	((c)->state & MMC_STATE_READONLY)
+#define mmc_card_blockaddr(c)	((c)->state & MMC_STATE_BLOCKADDR)
+#define mmc_card_ext_capacity(c) ((c)->state & MMC_CARD_SDXC)
+#define mmc_card_removed(c)	((c) && ((c)->state & MMC_CARD_REMOVED))
+#define mmc_card_doing_bkops(c)	((c)->state & MMC_STATE_DOING_BKOPS)
+
+#define mmc_sd_card_uhs(c)	((c)->state & MMC_STATE_ULTRAHIGHSPEED)
+
+#define mmc_card_suspended(c)  ((c)->state & MMC_STATE_SUSPENDED)
+
+#define mmc_card_set_present(c)	((c)->state |= MMC_STATE_PRESENT)
+#define mmc_card_set_readonly(c) ((c)->state |= MMC_STATE_READONLY)
+#define mmc_card_set_highspeed(c) ((c)->state |= MMC_STATE_HIGHSPEED)
+#define mmc_card_set_blockaddr(c) ((c)->state |= MMC_STATE_BLOCKADDR)
+#define mmc_card_set_ext_capacity(c) ((c)->state |= MMC_CARD_SDXC)
+#define mmc_card_set_removed(c) ((c)->state |= MMC_CARD_REMOVED)
+#define mmc_card_set_doing_bkops(c)	((c)->state |= MMC_STATE_DOING_BKOPS)
+#define mmc_card_clr_doing_bkops(c)	((c)->state &= ~MMC_STATE_DOING_BKOPS)
+#define mmc_card_set_suspended(c) ((c)->state |= MMC_STATE_SUSPENDED)
+#define mmc_card_clr_suspended(c) ((c)->state &= ~MMC_STATE_SUSPENDED)
+#define mmc_card_cmdq(c)   ((c)->state & MMC_STATE_CMDQ)
+#define mmc_card_set_cmdq(c)       ((c)->state |= MMC_STATE_CMDQ)
+#define mmc_card_clr_cmdq(c)       ((c)->state &= ~MMC_STATE_CMDQ)
+#define mmc_sd_card_set_uhs(c) ((c)->state |= MMC_STATE_ULTRAHIGHSPEED)
+
+#define mmc_card_name(c)	((c)->cid.prod_name)
+#define mmc_card_id(c)		(dev_name(&(c)->dev))
+
+#define mmc_dev_to_card(d)	container_of(d, struct mmc_card, dev)
+
+/*
+ *  The world is not perfect and supplies us with broken mmc/sdio devices.
+ *  For at least some of these bugs we need a work-around.
+ */
+#define CID_MANFID_SANDISK	0x2
+#define CID_MANFID_TOSHIBA	0x11
+#define CID_MANFID_MICRON	0x13
+#define CID_MANFID_SAMSUNG	0x15
+#define CID_MANFID_KINGSTON	0x70
+#define CID_MANFID_HYNIX	0x90
+
+#define CID_MANFID_ANY (-1u)
+#define CID_OEMID_ANY ((unsigned short) -1)
+#define CID_NAME_ANY (NULL)
+
+#define END_FIXUP { NULL }
+
+#define cid_rev(hwrev, fwrev, year, month)	\
+	(((u64) hwrev) << 40 |                  \
+	 ((u64) fwrev) << 32 |                  \
+	 ((u64) year) << 16 |                   \
+	 ((u64) month))
+
+#define cid_rev_card(card)		  \
+	cid_rev(card->cid.hwrev,	  \
+		    card->cid.fwrev,      \
+		    card->cid.year,	  \
+		    card->cid.month)
+#endif /* CONFIG_ASCEND_HISI_MMC */
 #endif /* LINUX_MMC_CARD_H */
diff --git a/include/linux/mmc/core.h b/include/linux/mmc/core.h
old mode 100644
new mode 100755
index 134a64833..c1283b973
--- a/include/linux/mmc/core.h
+++ b/include/linux/mmc/core.h
@@ -108,7 +108,9 @@ struct mmc_command {
  * ENOMEDIUM    Host can determine that the slot is empty and is
  *              actively failing requests
  */
-
+#ifdef CONFIG_ASCEND_HISI_MMC
+#define MMC_TIMEOUT_INVALID (0xFFFFFFFF)
+#endif
 	unsigned int		busy_timeout;	/* busy detect timeout in ms */
 	/* Set this flag only for blocking sanitize request */
 	bool			sanitize_busy;
diff --git a/include/linux/mmc/host.h b/include/linux/mmc/host.h
old mode 100644
new mode 100755
index 840462ed1..d8c320acf
--- a/include/linux/mmc/host.h
+++ b/include/linux/mmc/host.h
@@ -19,6 +19,9 @@
 #include <linux/mmc/pm.h>
 #include <linux/dma-direction.h>

+#include <linux/jiffies.h>
+#include <linux/version.h>
+
 struct mmc_ios {
 	unsigned int	clock;			/* clock rate */
 	unsigned short	vdd;
@@ -63,6 +66,7 @@ struct mmc_ios {
 #define MMC_TIMING_MMC_DDR52	8
 #define MMC_TIMING_MMC_HS200	9
 #define MMC_TIMING_MMC_HS400	10
+#define MMC_TIMING_NEW_SD    MMC_TIMING_UHS_SDR12

 	unsigned char	signal_voltage;		/* signalling voltage (1.8V or 3.3V) */

@@ -78,7 +82,25 @@ struct mmc_ios {
 #define MMC_SET_DRIVER_TYPE_D	3

 	bool enhanced_strobe;			/* hs400es selection */
+#ifdef CONFIG_ASCEND_HISI_MMC
+	unsigned int    clock_store;    /*store the clock before power off*/
+#endif
+};
+
+#ifdef CONFIG_ASCEND_HISI_MMC
+struct mmc_cmdq_host_ops {
+	int (*enable)(struct mmc_host *mmc);
+	int (*disable)(struct mmc_host *mmc, bool soft);
+	int (*restore_irqs)(struct mmc_host *mmc);
+	int (*request)(struct mmc_host *mmc, struct mmc_request *mrq);
+	int (*halt)(struct mmc_host *mmc, bool halt);
+	void (*post_req)(struct mmc_host *mmc, struct mmc_request *mrq,
+		int err);
+	void (*disable_immediately)(struct mmc_host *mmc);
+	int (*clear_and_halt)(struct mmc_host *mmc);
 };
+#endif
+

 struct mmc_host;

@@ -168,6 +190,12 @@ struct mmc_host_ops {
 	 */
 	int	(*multi_io_quirk)(struct mmc_card *card,
 				  unsigned int direction, int blk_size);
+#ifdef CONFIG_ASCEND_HISI_MMC
+    /* Slow down clk for ascend chip SD cards */
+	void (*slowdown_clk)(struct mmc_host *host, int timing);
+	int (*enable_enhanced_strobe)(struct mmc_host *host);
+	int (*send_cmd_direct)(struct mmc_host *host, struct mmc_request *mrq);
+#endif
 };

 struct mmc_cqe_ops {
@@ -255,6 +283,30 @@ struct mmc_context_info {
 	wait_queue_head_t	wait;
 };

+#ifdef CONFIG_ASCEND_HISI_MMC
+/**
+ * mmc_cmdq_context_info - describes the contexts of cmdq
+ * @active_reqs		requests being processed
+ * @active_dcmd		dcmd in progress, don't issue any
+ *			more dcmd requests
+ * @rpmb_in_wait	do not pull any more reqs till rpmb is handled
+ * @cmdq_state		state of cmdq engine
+ * @req_starved		completion should invoke the request_fn since
+ *			no tags were available
+ * @cmdq_ctx_lock	acquire this before accessing this structure
+ */
+struct mmc_cmdq_context_info {
+	unsigned long	active_reqs; /* in-flight requests */
+	bool		active_dcmd;
+	bool		rpmb_in_wait;
+	unsigned long   curr_state;
+
+	/* no free tag available */
+	unsigned long	req_starved;
+	spinlock_t	cmdq_ctx_lock;
+};
+#endif
+
 struct regulator;
 struct mmc_pwrseq;

@@ -328,10 +380,14 @@ struct mmc_host {
 #define MMC_CAP_UHS_SDR50	(1 << 18)	/* Host supports UHS SDR50 mode */
 #define MMC_CAP_UHS_SDR104	(1 << 19)	/* Host supports UHS SDR104 mode */
 #define MMC_CAP_UHS_DDR50	(1 << 20)	/* Host supports UHS DDR50 mode */
+#ifdef CONFIG_ASCEND_HISI_MMC
+#define MMC_CAP_RUNTIME_RESUME	(1 << 20)	/* Resume at runtime_resume. */
+#endif
 #define MMC_CAP_UHS		(MMC_CAP_UHS_SDR12 | MMC_CAP_UHS_SDR25 | \
 				 MMC_CAP_UHS_SDR50 | MMC_CAP_UHS_SDR104 | \
 				 MMC_CAP_UHS_DDR50)
 /* (1 << 21) is free for reuse */
+#define MMC_CAP_NEED_RSP_BUSY	(1 << 22)	/* Commands with R1B can't use R1. */
 #define MMC_CAP_DRIVER_TYPE_A	(1 << 23)	/* Host supports Driver Type A */
 #define MMC_CAP_DRIVER_TYPE_C	(1 << 24)	/* Host supports Driver Type C */
 #define MMC_CAP_DRIVER_TYPE_D	(1 << 25)	/* Host supports Driver Type D */
@@ -367,6 +423,34 @@ struct mmc_host {
 #define MMC_CAP2_CQE		(1 << 23)	/* Has eMMC command queue engine */
 #define MMC_CAP2_CQE_DCMD	(1 << 24)	/* CQE can issue a direct command */
 #define MMC_CAP2_AVOID_3_3V	(1 << 25)	/* Host must negotiate down from 3.3V */
+#ifdef CONFIG_ASCEND_HISI_MMC
+#define MMC_CAP2_CACHE_CTRL	(1 << 1)	/* Allow cache control */
+#define MMC_CAP2_NO_MULTI_READ	(1 << 3)	/* Multiblock read don't work */
+#define MMC_CAP2_NO_SLEEP_CMD	(1 << 4)	/* Don't allow sleep command */
+#define MMC_CAP2_BROKEN_VOLTAGE	(1 << 7)	/* Use the broken voltage */
+#define MMC_CAP2_DETECT_ON_ERR	(1 << 8)	/* I/O err check card removal */
+#define MMC_CAP2_HC_ERASE_SZ	(1 << 9)	/* High-capacity erase size */
+#define MMC_CAP2_PACKED_RD	(1 << 12)	/* Allow packed read */
+#define MMC_CAP2_PACKED_WR	(1 << 13)	/* Allow packed write */
+#define MMC_CAP2_PACKED_CMD	(MMC_CAP2_PACKED_RD | \
+				 MMC_CAP2_PACKED_WR)
+#define MMC_CAP2_CMD_QUEUE	(1 << 18)	/* support eMMC command queue */
+#define MMC_CAP2_ENHANCED_STROBE	(1 << 19)
+#define MMC_CAP2_CACHE_FLUSH_BARRIER	(1 << 20)
+/* Allow background operations auto enable control */
+#define MMC_CAP2_BKOPS_AUTO_CTRL	(1 << 21)
+/* Allow background operations manual enable control */
+#define MMC_CAP2_BKOPS_MANUAL_CTRL	(1 << 22)
+
+/* host is connected by via modem through sdio */
+#define MMC_CAP2_SUPPORT_VIA_MODEM		(1 << 26)
+/* host is connected by wifi through sdio */
+#define MMC_CAP2_SUPPORT_WIFI			(1 << 27)
+/* host is connected to 1102 wifi */
+#define MMC_CAP2_SUPPORT_WIFI_CMD11		(1 << 28)
+/* host do not support low power for wifi*/
+#define MMC_CAP2_WIFI_NO_LOWPWR		(1 << 29)
+#endif

 	int			fixed_drv_type;	/* fixed driver type for non-removable media */

@@ -459,6 +543,12 @@ struct mmc_host {
 	bool			cqe_enabled;
 	bool			cqe_on;

+#ifdef CONFIG_ASCEND_HISI_MMC
+	const struct mmc_cmdq_host_ops *cmdq_ops;
+	int					sdio_present;
+	unsigned int    cmdq_slots;
+	struct mmc_cmdq_context_info    cmdq_ctx;
+#endif
 	unsigned long		private[0] ____cacheline_aligned;
 };

@@ -587,4 +677,30 @@ static inline enum dma_data_direction mmc_get_dma_dir(struct mmc_data *data)
 int mmc_send_tuning(struct mmc_host *host, u32 opcode, int *cmd_error);
 int mmc_abort_tuning(struct mmc_host *host, u32 opcode);

+#ifdef CONFIG_ASCEND_HISI_MMC
+int mmc_cache_ctrl(struct mmc_host *host, u8 enable);
+int mmc_card_awake(struct mmc_host *host);
+int mmc_card_sleep(struct mmc_host *host);
+int mmc_card_can_sleep(struct mmc_host *host);
+#else
+static inline int mmc_cache_ctrl(struct mmc_host *host, u8 enable)
+{
+	return 0;
+}
+static inline int mmc_card_awake(struct mmc_host *host)
+{
+	return 0;
+}
+static inline int mmc_card_sleep(struct mmc_host *host)
+{
+	return 0;
+}
+static inline int mmc_card_can_sleep(struct mmc_host *host)
+{
+	return 0;
+}
+#endif
+
+int mmc_is_ascend_customized(struct device *dev);
+
 #endif /* LINUX_MMC_HOST_H */
diff --git a/include/linux/mmc/mmc.h b/include/linux/mmc/mmc.h
old mode 100644
new mode 100755
index 897a87c4c..ab082f65a
--- a/include/linux/mmc/mmc.h
+++ b/include/linux/mmc/mmc.h
@@ -387,6 +387,52 @@ static inline bool mmc_op_multi(u32 opcode)

 #define EXT_CSD_PACKED_EVENT_EN	BIT(3)

+#ifdef CONFIG_ASCEND_HISI_MMC
+#define MMC_FLUSH_REQ_TIMEOUT_MS 90000 /* msec */
+
+/* EXT_CSD field definitions for hisi mmc */
+#define EXT_CSD_CMDQ_MODE		15	/* R/W */
+#define EXT_CSD_FFU_STATUS		26	/* R */
+#define EXT_CSD_MODE_OPERATION_CODES	29	/* W */
+#define EXT_CSD_MODE_CONFIG		30	/* R/W */
+#define EXT_CSD_BARRIER_CTRL		31	/* W */
+#define EXT_CSD_VENDOR_FEATURE_SUPPORT	124	/* Reserved R */
+#define EXT_CSD_USER_WP			171	/* R/W */
+#define EXT_CSD_RPMB_REGION1_SIZE_MULT	180	/* Reserved R/W */
+#define EXT_CSD_RPMB_REGION2_SIZE_MULT	182	/* Reserved R/W */
+#define EXT_CSD_RPMB_REGION3_SIZE_MULT	186	/* Reserved R/W */
+#define EXT_CSD_RPMB_REGION4_SIZE_MULT	188	/* Reserved R/W */
+#define EXT_CSD_SLEEP_NOTIFICATION_TIME 216	/* Ro */
+#define EXT_CSD_CACHE_FLUSH_POLICY	240	/* RO */
+#define EXT_CSD_HEALTH_REPORT		270	/* RO 32 bits*/
+#define EXT_CSD_NUM_OF_FW_SEC_PROG	302	/* RO */
+#define EXT_CSD_BARRIER_SUPPORT		486	/* RO */
+#define EXT_CSD_FFU_ARG			487	/* RO, 4 bytes */
+#define EXT_CSD_OPERATION_CODE_TIMEOUT	491	/* RO */
+#define EXT_CSD_FFU_FEATURES		492	/* RO */
+#define EXT_CSD_RPMB_REL_WR_EN			(1<<4)
+#define EXT_CSD_SLEEP_NOTIFICATION	4
+
+#define EXT_CSD_TIMING_BC	0	/* Backwards compatility */
+#define EXT_CSD_TIMING_HS	1	/* High speed */
+#define EXT_CSD_TIMING_HS200	2	/* HS200 */
+#define EXT_CSD_TIMING_HS400	3	/* HS400 */
+
+#define EXT_CSD_DRVIER_STRENGTH_SHIFT	4 /*Hight 4bit of EXT_CSD_HS_TIMING*/
+#define EXT_CSD_DRVIER_STRENGTH_50		0x0
+#define EXT_CSD_DRVIER_STRENGTH_33		0x1
+#define EXT_CSD_DRVIER_STRENGTH_66		0x2
+#define EXT_CSD_DRVIER_STRENGTH_100		0x3
+#define EXT_CSD_DRVIER_STRENGTH_40		0x4
+
+#define EXT_CSD_STROBE_ENHANCED			BIT(7)
+#define EXT_CSD_TRIGGER_CACHE_FLUSH		BIT(0)
+#define EXT_CSD_SET_CACHE_FLUSH_BARRIER		BIT(1)
+#define EXT_CSD_CACHE_FLUSH_POLICY_FIFO		BIT(0)
+#define EXT_CSD_BKOPS_MANUAL_EN			BIT(0)
+#define EXT_CSD_BKOPS_AUTO_EN			BIT(1)
+#endif /* CONFIG_ASCEND_HISI_MMC */
+
 /*
  * EXCEPTION_EVENT_STATUS field
  */
diff --git a/include/linux/mmc/pm.h b/include/linux/mmc/pm.h
old mode 100644
new mode 100755
index 4a139204c..6e2d6a135
--- a/include/linux/mmc/pm.h
+++ b/include/linux/mmc/pm.h
@@ -26,5 +26,6 @@ typedef unsigned int mmc_pm_flag_t;

 #define MMC_PM_KEEP_POWER	(1 << 0)	/* preserve card power during suspend */
 #define MMC_PM_WAKE_SDIO_IRQ	(1 << 1)	/* wake up host system on SDIO IRQ assertion */
+#define MMC_PM_IGNORE_PM_NOTIFY	(1 << 2)	/* ignore mmc pm notify */

 #endif /* LINUX_MMC_PM_H */
--
2.17.1

